fastavro/_write_py.py

```diff
@@ -296,89 +312,174 @@ def write_map(fo, datum, schema):
                 for f in schema['fields']
             )
-        )
+    )
 
-    if record_type in SCHEMA_DEFS:
-        return validate(datum, SCHEMA_DEFS[record_type])
 
-    raise ValueError('unkown record type - %s' % record_type)
+def validate_union(datum, schema, raise_errors=False):
+    if isinstance(datum, tuple):
+        (name, datum) = datum
+        for candidate in schema:
+            if extract_record_type(candidate) == 'record':
+                if name == candidate["name"]:
+                    return validate(datum, schema=candidate,
+                                    field=None,
+                                    raise_errors=raise_errors)
+        else:
+            return False
+
+    errors = []
+    for s in schema:
+        try:
+            ret = validate(datum, schema=s,
+                           field=None,
+                           raise_errors=raise_errors)
+            if ret:
+                # We exit on the first passing type in Unions
+                return True
+        except ValidationException as e:
+            errors.append(e.message)
+    if raise_errors:
+        raise ValidationErrors(errors)
+    return False
+
+
+VALIDATORS = {
+    'null': validate_null,
+    'boolean': validate_boolean,
+    'string': validate_string,
+    'int': validate_int,
+    'long': validate_long,
+    'float': validate_float,
+    'double': validate_float,
+    'bytes': validate_bytes,
+    'fixed': validate_fixed,
+    'enum': validate_enum,
+    'array': validate_array,
+    'map': validate_map,
+    'union': validate_union,
+    'error_union': validate_union,
+    'record': validate_record,
+    'error': validate_record,
+    'request': validate_record
+}
+
+
+def validate(datum, schema, field=None, raise_errors=False):
+    """Determine if a python datum is an instance of a schema."""
+    record_type = extract_record_type(schema)
+    result = None
+
+    if hasattr(schema, 'get'):
+        if field is not None:
+            ns_field = '.'.join([field, schema.get('name')])
+        else:
+            ns_field = schema.get('name')
+    else:
+        ns_field = field
+
+    if record_type in ('null', 'union'):
+        validator = VALIDATORS.get(record_type)
+        result = validator(datum, schema=schema, raise_errors=raise_errors)
+    elif datum is None:
+        # data required if not Null type
+        result = False
+    else:
+        validator = VALIDATORS.get(record_type)
+        if validator:
+            result = validator(datum, schema=schema, raise_errors=raise_errors)
+
+    if record_type in SCHEMA_DEFS and result is None:
+        result = validate(datum,
+                          schema=SCHEMA_DEFS[record_type],
+                          field=ns_field,
+                          raise_errors=raise_errors)
+
+    if raise_errors and result is False:
+        raise ValidationException(datum, schema, ns_field)
+
+    if result is None:
+        raise UnknownType(record_type)
```

Comment:
This is technically a breaking change as `validate` used to throw a `ValueError` when there was an unknown schema, but I like this better and I don't think there would be much to any impact. :+1:

Reply:
Ya when I saw the `UnknownType` in schema, I was assuming it was meant for here but never migrated for use.

Reply:
I have a fix for the "breaking change", we update UnknownType to inherit `ValueError`.

```
class UnknownType(ValueError):
```

I've made the change and tested in main, would you like me to create any unittest around it?